internal/metrics/metrics.go

```diff
@@ -0,0 +1,199 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"git.sr.ht/~spc/go-log"
+	"github.com/jakub-dzon/k4e-operator/models"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"os"
+	"path"
+	"sync"
+	"time"
+)
+
+const (
+	// DefaultRetentionDuration specifies 7 days data retention
+	DefaultRetentionDuration = 7 * 24 * time.Hour
+	// DefaultMaxBytes specifies no block data size limit
+	DefaultMaxBytes = int64(0)
+)
+
+type DataPoint struct {
+	Time  int64
+	Value float64
+}
+type Series struct {
+	Labels     map[string]string
+	DataPoints []DataPoint
+}
+
+//go:generate mockgen -package=metrics -destination=storage_mock.go . Storage
+type Storage interface {
+	Deregister() error
+	GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error)
+	AddMetric(value float64, labelsMap map[string]string) error
+}
+
+type TSDB struct {
+	appliedOptions *tsdb.Options
+	db             *tsdb.DB
+	dbLock         sync.RWMutex
+}
+
+func NewTSDB(dataDir string) (*TSDB, error) {
+	metricsDir := path.Join(dataDir, "metrics")
+	if err := os.MkdirAll(metricsDir, 0755); err != nil {
+		return nil, fmt.Errorf("cannot create directory: %w", err)
+	}
+	opts := tsdb.DefaultOptions()
+
+	opts.RetentionDuration = int64(DefaultRetentionDuration / time.Millisecond)
+	opts.MaxBytes = DefaultMaxBytes
+
+	db, err := tsdb.Open(metricsDir, nil, nil, opts, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &TSDB{
+		db:             db,
+		dbLock:         sync.RWMutex{},
+		appliedOptions: opts,
+	}, err
+}
+
+func (t *TSDB) Deregister() error {
+	metricsDir := t.db.Dir()
+	t.dbLock.Lock()
+	defer t.dbLock.Unlock()
+
+	err := t.db.Close()
```

Comment:
shouldn't this be called every time worker stops/terminates/shutsdown?

Reply:
Fixed - added signal handler in main.